src/ingameWindows/iwMapGenerator.cpp

```diff
@@ -0,0 +1,313 @@
+{
+    AddTextButton(0,  20, 360, 100, 20, TC_RED2, _("Back"), NormalFont);
+    AddTextButton(1, 130, 360, 100, 20, TC_GREEN2, _("Apply"), NormalFont);
+
+    ctrlComboBox* combo = AddComboBox(2, 20, 30, 210, 20, TC_GREY, NormalFont, 100);
+    combo->AddString("2 Players");
+    combo->AddString("3 Players");
+    combo->AddString("4 Players");
+    combo->AddString("5 Players");
+    combo->AddString("6 Players");
+    combo->AddString("7 Players");
+    
+    combo = AddComboBox(3, 20, 60, 210, 20, TC_GREY, NormalFont, 100);
+    combo->AddString("Islands");
+    combo->AddString("Continent");
+    combo->AddString("Greenland");
+    combo->AddString("Migration");
+    combo->AddString("Riverland");
+    combo->AddString("Ringland");
+    combo->AddString("Random");
+
+    combo = AddComboBox(4, 20, 90, 210, 20, TC_GREY, NormalFont, 100);
+    combo->AddString("64 x 64");
+    combo->AddString("128 x 128");
+    combo->AddString("256 x 256");
+    combo->AddString("512 x 512");
+    combo->AddString("1024 x 1024");
+
+    AddText(5, 20, 120, _("Player Distribution"), COLOR_YELLOW, 0, NormalFont);
+    combo = AddComboBox(6, 20, 140, 210, 20, TC_GREY, NormalFont, 100);
+    combo->AddString("Very Close");
+    combo->AddString("Close");
+    combo->AddString("Medium");
+    combo->AddString("Far");
+    combo->AddString("Very Far");
+
+    AddText(7, 20, 170, _("Landscape"), COLOR_YELLOW, 0, NormalFont);
+    combo = AddComboBox(8, 20, 190, 210, 20, TC_GREY, NormalFont, 100);
+    combo->AddString("Greenland");
+    combo->AddString("Winterworld");
+    combo->AddString("Wasteland");
+    
+    AddText(9, 20, 225, _("Gold:"), COLOR_YELLOW, 0, NormalFont);
+    ctrlProgress* progressBar = AddProgress(10, 100, 220, 130, 20,
+                                            TC_GREY, 139, 138, 100);
+    progressBar->SetPosition(settings.ratioGold);
+
+    AddText(11, 20, 255, _("Iron:"), COLOR_YELLOW, 0, NormalFont);
+    progressBar = AddProgress(12, 100, 250, 130, 20,
+                              TC_GREY, 139, 138, 100);
+    progressBar->SetPosition(settings.ratioIron);
+
+    AddText(13, 20, 285, _("Coal:"), COLOR_YELLOW, 0, NormalFont);
+    progressBar = AddProgress(14, 100, 280, 130, 20,
+                              TC_GREY, 139, 138, 100);
+    progressBar->SetPosition(settings.ratioCoal);
+
+    AddText(15, 20, 315, _("Granite:"), COLOR_YELLOW, 0, NormalFont);
+    progressBar = AddProgress(16, 100, 310, 130, 20,
+                              TC_GREY, 139, 138, 100);
+    progressBar->SetPosition(settings.ratioGranite);
+    
+    Reset();
+}
+
+iwMapGenerator::~iwMapGenerator()
+{
+}
+
+void iwMapGenerator::Msg_ButtonClick(const unsigned int ctrl_id)
+{
+    switch(ctrl_id)
+    {
+        default:
+            break;
+
+        case 0: // back
+        {
+            Close();
+        } break;
+            
+        case 1: // apply
+        {
+            tmpSettings.ratioGold = GetCtrl<ctrlProgress>(10)->GetPosition();
```

Comment:
Why not get rid of tmpSettings and query all the values here? No need to react on the change events.